cmd/bad-key-revoker/main.go

```diff
@@ -471,19 +477,85 @@ func main() {
 		logger:          logger,
 		clk:             clk,
 	}
+
+	// If the MaxBackoff.Duration was not set via the config, set it now
+	// so that it can't increase so high that it will not try on a
+	// reasonable timeframe.
+	if config.BadKeyRevoker.MaxBackoff.Duration == 0 {
+		config.BadKeyRevoker.MaxBackoff.Duration = time.Second * 60
+	}
+
+	// Set the backup policy
+	backoff := &BackoffPolicy{
+		Limit:   config.BadKeyRevoker.MaxBackoff.Duration,
+		Minimum: config.BadKeyRevoker.Interval.Duration,
+		Factor:  1.3,
+	}
+
+	// Run bad key revoker in a loop. Backoff if no work or errors. Increment error or backoff metrics.
 	for {
 		noWork, err := bkr.invoke()
 		if err != nil {
 			keysProcessed.WithLabelValues("error").Inc()
 			logger.AuditErrf("failed to process blockedKeys row: %s", err)
+			// Backoff on error
+			backoff.increase()
+			logger.Infof("error. Trying again in %f seconds. Backoff counter: %d", backoff.Value.Seconds(), backoff.Counter)
 			continue
 		}
 		if noWork {
-			logger.Info(fmt.Sprintf(
-				"No work to do. Sleeping for %s", config.BadKeyRevoker.Interval.Duration))
-			time.Sleep(config.BadKeyRevoker.Interval.Duration)
+			// Backoff on no work to do
+			backoff.increase()
+			logger.Infof("No work to do. Trying again in %f seconds. Backoff counter: %d", backoff.Value.Seconds(), backoff.Counter)
 		} else {
 			keysProcessed.WithLabelValues("success").Inc()
+			// Successfully processed. Reset backoff.
+			backoff.reset()
 		}
+		// Sleep for duration defined in the backoff policy
+		bkr.clk.Sleep(backoff.Value)
+	}
+}
+
+// Policy for backing off requests for BadKeyRevoker runs. BadKeyRevoker
+// doesn't use goroutines, so we don't need jitter to stagger
+// concurrent retries.
+type BackoffPolicy struct {
+	Limit   time.Duration // Maximum backoff time
+	Value   time.Duration // Current backoff time
+	Counter int64         // Backoff counter
+	Minimum time.Duration // Minimum time to wait to do work
+	Factor  float64       // Backoff factor
```

Comment:
For exported fields, comments should go above the field they reference and be in the form of a complete sentence that begins with the field name.

```suggestion
        // Limit is the maximum backoff time.
	Limit   time.Duration
	...
```

Reply:
Good point. I don't think any of this needs to be exported, including the struct. I think I will just change them all to local.

Reply:
No longer relevant for this pr. I did away with the struct.